docs/cudf/source/developer_guide/developer_guide.md

```diff
@@ -0,0 +1,130 @@
+# Developer Guide
+
+```{note}
+At present, this guide only covers the main cuDF library.
+In the future, it may be expanded to also cover dask_cudf, cudf_kafka, and custreamz.
+```
+
+The cuDF library is a GPU-accelerated, [Pandas-like](https://pandas.pydata.org/) DataFrame library.
+Under the hood, all of cuDF's functionality relies on the CUDA-accelerated `libcudf` C++ library.
+Thus, cuDF's internals are designed to efficiently and robustly map pandas APIs to `libcudf` functions.
+For more information about the `libcudf` library, a good starting point is the
+[developer guide](https://github.com/rapidsai/cudf/blob/main/cpp/docs/DEVELOPER_GUIDE.md).
+
+This document assumes familiarity with the
+[overall contributing guide](https://github.com/rapidsai/cudf/blob/main/CONTRIBUTING.md).
+The goal of this document is to provide more specific guidance for Python developers.
+It covers the structure of the Python code and discusses best practices.
+Additionally, it includes longer sections on more specific topics like testing and benchmarking.
+More specific information on these can be found in the pages below:
+
+```{toctree}
+:maxdepth: 1
+
+library_design
+```
+
+The rest of this document focuses on a higher-level overview of best practices in cuDF.
+
+## Directory structure and file naming
+
+cuDF generally presents the same importable modules and subpackages as pandas.
+All Cython code is contained in `python/cudf/cudf/_lib`.
+
+**Open question**: Should we start enforcing a stricter file/directory layout? Any suggestions?
+
+
+## Code style
+
+cuDF employs a number of linters to ensure consistent style across the code base:
+
+- [`flake8`](https://github.com/pycqa/flake8) checks for general code formatting compliance. 
+- [`black`](https://github.com/psf/black) is an automatic code formatter.
+  Code formatted with `black` is largely `flake8`-compliant, although there are a few minor issues that it cannot fix automatically.
+- [`isort`](https://pycqa.github.io/isort/) ensures imports are sorted consistently.
+- [`mypy`](http://mypy-lang.org/) performs static type checking.
+  In conjunction with [type hints](https://docs.python.org/3/library/typing.html),
+  `mypy` can help catch various bugs that are otherwise difficult to find.
+- [`pydocstyle`](https://github.com/PyCQA/pydocstyle/) lints docstring style.
+
+Configuration information for these tools is contained in a number of files.
+The primary source of truth is the `.pre-commit-config.yaml` file at the root of the repo.
+As described in 
+[overall contributing guide](https://github.com/rapidsai/cudf/blob/main/CONTRIBUTING.md),
+we recommend using [`pre-commit`](https://pre-commit.com/) to manage all linters.
+
+## Deprecating and removing code
+
+cuDF generally follows the policy of deprecating code for one release prior to removal.
+For example, if we want to remove an API after the 22.06 release, it will be marked as deprecated in the 22.08 release.
```

Comment:
The mention of 22.06 is confusing. Maybe “if we decide to remove an API during the 22.08 development cycle, it will be deprecated in the 22.08 release and removed in the 22.10 release.”